[JSC] Move Promise.resolve and Promise.reject to C++

Constellation opened the pull request:
#### a0faddd38a6d4f20735b1404c116e79eee972275
<pre>
[JSC] Move Promise.resolve and Promise.reject to C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=300603">https://bugs.webkit.org/show_bug.cgi?id=300603</a>
<a href="https://rdar.apple.com/162500561">rdar://162500561</a>

Reviewed by Yijia Huang.

This patch moves Promise.resolve / Promise.reject implementations from
JS to C++ to improve stability and performance. Doing so will further
allow us to move more things (e.g. Promise.all) to C++ to gain
significant performance improvement.

The most difficult part of this is Promise.resolve and its optimization.
Since we were using inlined Promise.resolve and avoiding various costly
operations when we detect some input values in DFG / FTL, we cannot just
move them to C++. In this patch, we implement intrinsic for them and add
DFG nodes handling these optimizations.

1. We check Promise input and if it is fine &amp; watchpoint is fine, then
   we will just convert Promise.resolve(promise) to identity.
2. We check non object input and convert it to fulfilled promise
   allocation.
3. We check an object input without having &quot;then&quot; property and convert
   it to (2) as well with watchpoints.

Also, we expose @promiseResolve / @promiseReject abstract operations
and use tme instead of Promise.@resolve / Promise.@reject.

* JSTests/stress/dfg-promise-resolve-fulfilled-identity.js: Added.
(shouldBe):
(test):
(async testMain):
* JSTests/stress/dfg-promise-resolve-fulfilled-then.js: Added.
(shouldBe):
(test):
(DerivedPromise.prototype.then):
(DerivedPromise):
(async testMain):
* JSTests/stress/dfg-promise-resolve-fulfilled.js: Added.
(shouldBe):
(test):
(async testMain):
* Source/JavaScriptCore/builtins/AsyncDisposableStackPrototype.js:
(disposeAsync):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DisposableStackPrototype.js:
(linkTimeConstant.getAsyncDisposableMethod):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseNewOnRejected):
(linkTimeConstant.promiseNewOnRejected.return.reject): Deleted.
(reject): Deleted.
(resolve): Deleted.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.newPromiseCapability):
(linkTimeConstant.newPromiseCapability.resolve): Deleted.
(linkTimeConstant.newPromiseCapability.reject): Deleted.
(linkTimeConstant.promiseResolve): Deleted.
(linkTimeConstant.promiseResolveSlow): Deleted.
(linkTimeConstant.promiseRejectSlow): Deleted.
* Source/JavaScriptCore/builtins/WebAssembly.js:
(compileStreaming):
(instantiateStreaming):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
(JSC::DFG::JITData::tryInitialize):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::promiseResolve):
(JSC::JSPromise::promiseReject):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::create):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromiseConstructor::addOwnInternalSlots): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js:
(cancel):
(read):
(getter.closed):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(readableStreamPipeToForBindings):
(readableStreamDefaultReaderClosedForBindings):
(readableStreamDefaultReaderReadForBindings):
(readableStreamDefaultReaderCancelForBindings):
(readableStreamReaderGenericInitialize):
(readableStreamPipeToWritableStream):
(const.pullAlgorithm):
(readableStreamTeePullFunction):
(readableStreamCancel):
(readableStreamClose):
(readableStreamFulfillReadRequest):
(readableStreamDefaultReaderRead):
(readableStreamDefaultReaderErrorReadRequests):
* Source/WebCore/Modules/streams/StreamInternals.js:
(shieldingPromiseResolve):
(promiseInvokeOrNoopMethodNoCatch):
(promiseInvokeOrNoopMethod):
(promiseInvokeOrNoop):
(promiseInvokeOrFallbackOrNoop):
(createFulfilledPromise):
* Source/WebCore/Modules/streams/TransformStreamInternals.js:
(initializeTransformStream):
(setUpTransformStreamDefaultControllerFromTransformer):
(transformStreamDefaultSinkAbortAlgorithm):
* Source/WebCore/Modules/streams/WritableStreamDefaultWriter.js:
(getter.closed):
(getter.ready):
(abort):
(close):
(write):
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(writableStreamCloseForBindings):
(writableStreamAbortForBindings):
(writableStreamAbort):
(writableStreamClose):
(writableStreamDefaultWriterCloseWithErrorPropagation):
(writableStreamDefaultWriterWrite):
(setUpWritableStreamDefaultController):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::whenPromiseIsSettled):
* Source/WebCore/bindings/js/JSDOMPromise.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.cpp:
(WebCore::createRejectedPromiseWithTypeError):
* Source/WebCore/dom/TextDecoderStream.js:
(initializeTextDecoderStream):
* Source/WebCore/dom/TextEncoderStream.js:
(initializeTextEncoderStream):

Canonical link: <a href="https://commits.webkit.org/301423@main">https://commits.webkit.org/301423@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70f215f89230ce2a208069fd3b94ab312066267b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125898 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45560 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36315 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132768 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77762 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/506c1007-d3e9-4ead-b9ca-6fd3b516351f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127769 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54118 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95920 "") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64017 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/60758679-a83b-41e4-bf93-44dcdc2a8d1c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36982 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112582 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76411 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30766 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76240 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117978 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135451 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/124405 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52683 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40422 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104385 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53132 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104112 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49484 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27800 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50054 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19708 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52577 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58389 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/157418 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51920 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39417 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55271 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53619 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -124,7 +124,7 @@ std::tuple<JSObject*, JSObject*, JSObject*> JSPromise::newPromiseCapability(JSGl
     }
 
-    auto* executor = JSFunctionWithFields::create(vm, globalObject, vm.promiseCapabilityExecutorExecutable(), 2, nullString());
+    auto* executor = JSFunctionWithFields::create(vm, globalObject, vm.promiseCapabilityExecutorExecutable(), 2, emptyString());
```
Is this change correct here? Since the others still use `nullString()`?
```
auto* resolve = JSFunctionWithFields::create(..., nullString());
auto* reject = JSFunctionWithFields::create(..., nullString());
```

Constellation replied:
Right now, we are working around the issue with this function since this executor needs `""` name actually. So we are using emptyString here instead of `nullString`. We would like to enhance `JSFunction` to support things well better way.

hyjorc1 reviewed (approved):
r=me with one comment

Constellation committed:
[JSC] Move Promise.resolve and Promise.reject to C++
https://bugs.webkit.org/show_bug.cgi?id=300603
rdar://162500561

Reviewed by Yijia Huang.

This patch moves Promise.resolve / Promise.reject implementations from
JS to C++ to improve stability and performance. Doing so will further
allow us to move more things (e.g. Promise.all) to C++ to gain
significant performance improvement.

The most difficult part of this is Promise.resolve and its optimization.
Since we were using inlined Promise.resolve and avoiding various costly
operations when we detect some input values in DFG / FTL, we cannot just
move them to C++. In this patch, we implement intrinsic for them and add
DFG nodes handling these optimizations.

1. We check Promise input and if it is fine & watchpoint is fine, then
   we will just convert Promise.resolve(promise) to identity.
2. We check non object input and convert it to fulfilled promise
   allocation.
3. We check an object input without having "then" property and convert
   it to (2) as well with watchpoints.

Also, we expose @promiseResolve / @promiseReject abstract operations
and use tme instead of Promise.@resolve / Promise.@reject.

* JSTests/stress/dfg-promise-resolve-fulfilled-identity.js: Added.
(shouldBe):
(test):
(async testMain):
* JSTests/stress/dfg-promise-resolve-fulfilled-then.js: Added.
(shouldBe):
(test):
(DerivedPromise.prototype.then):
(DerivedPromise):
(async testMain):
* JSTests/stress/dfg-promise-resolve-fulfilled.js: Added.
(shouldBe):
(test):
(async testMain):
* Source/JavaScriptCore/builtins/AsyncDisposableStackPrototype.js:
(disposeAsync):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/DisposableStackPrototype.js:
(linkTimeConstant.getAsyncDisposableMethod):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(linkTimeConstant.promiseNewOnRejected):
(linkTimeConstant.promiseNewOnRejected.return.reject): Deleted.
(reject): Deleted.
(resolve): Deleted.
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.newPromiseCapability):
(linkTimeConstant.newPromiseCapability.resolve): Deleted.
(linkTimeConstant.newPromiseCapability.reject): Deleted.
(linkTimeConstant.promiseResolve): Deleted.
(linkTimeConstant.promiseResolveSlow): Deleted.
(linkTimeConstant.promiseRejectSlow): Deleted.
* Source/JavaScriptCore/builtins/WebAssembly.js:
(compileStreaming):
(instantiateStreaming):
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGConstantFoldingPhase.cpp:
(JSC::DFG::ConstantFoldingPhase::foldConstants):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGJITCode.cpp:
(JSC::DFG::JITData::JITData):
(JSC::DFG::JITData::tryInitialize):
* Source/JavaScriptCore/dfg/DFGJITCode.h:
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::promiseResolve):
(JSC::JSPromise::promiseReject):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::create):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSPromiseConstructor::addOwnInternalSlots): Deleted.
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/WebCore/Modules/compression/CompressionStream.js:
(initializeCompressionStream):
* Source/WebCore/Modules/compression/DecompressionStream.js:
(initializeDecompressionStream):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js:
(cancel):
(read):
(getter.closed):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(readableStreamPipeToForBindings):
(readableStreamDefaultReaderClosedForBindings):
(readableStreamDefaultReaderReadForBindings):
(readableStreamDefaultReaderCancelForBindings):
(readableStreamReaderGenericInitialize):
(readableStreamPipeToWritableStream):
(const.pullAlgorithm):
(readableStreamTeePullFunction):
(readableStreamCancel):
(readableStreamClose):
(readableStreamFulfillReadRequest):
(readableStreamDefaultReaderRead):
(readableStreamDefaultReaderErrorReadRequests):
* Source/WebCore/Modules/streams/StreamInternals.js:
(shieldingPromiseResolve):
(promiseInvokeOrNoopMethodNoCatch):
(promiseInvokeOrNoopMethod):
(promiseInvokeOrNoop):
(promiseInvokeOrFallbackOrNoop):
(createFulfilledPromise):
* Source/WebCore/Modules/streams/TransformStreamInternals.js:
(initializeTransformStream):
(setUpTransformStreamDefaultControllerFromTransformer):
(transformStreamDefaultSinkAbortAlgorithm):
* Source/WebCore/Modules/streams/WritableStreamDefaultWriter.js:
(getter.closed):
(getter.ready):
(abort):
(close):
(write):
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(writableStreamCloseForBindings):
(writableStreamAbortForBindings):
(writableStreamAbort):
(writableStreamClose):
(writableStreamDefaultWriterCloseWithErrorPropagation):
(writableStreamDefaultWriterWrite):
(setUpWritableStreamDefaultController):
* Source/WebCore/bindings/js/JSDOMPromise.cpp:
(WebCore::DOMPromise::whenPromiseIsSettled):
* Source/WebCore/bindings/js/JSDOMPromise.h:
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.cpp:
(WebCore::createRejectedPromiseWithTypeError):
* Source/WebCore/dom/TextDecoderStream.js:
(initializeTextDecoderStream):
* Source/WebCore/dom/TextEncoderStream.js:
(initializeTextEncoderStream):

Canonical link: https://commits.webkit.org/301423@main
